class.two-factor-force.php

```diff
@@ -0,0 +1,488 @@
+		// Forced 2fa login functionality.
+		add_action( 'init', array( __CLASS__, 'register_scripts' ) );
+		add_action( 'wp_ajax_two_factor_force_form_submit', array( __CLASS__, 'handle_force_2fa_submission' ) );
+
+		// Handling intercession in 2 separate hooks to allow us to properly parse for REST requests.
+		add_filter( 'login_redirect', array( __CLASS__, 'maybe_redirect_login' ), 15, 3 );
+		add_action( 'parse_request', array( __CLASS__, 'maybe_redirect_to_2fa_settings' ) );
+		add_action( 'admin_init', array( __CLASS__, 'maybe_redirect_to_2fa_settings' ) );
+		add_action( 'admin_init', array( __CLASS__, 'maybe_display_2fa_settings' ) );
+
+		if ( is_multisite() ) {
+			add_action( 'wpmu_options', array( __CLASS__, 'force_two_factor_setting_options' ) );
+			add_action( 'update_wpmu_options', array( __CLASS__, 'save_network_force_two_factor_update' ) );
+		} else {
+			add_action( 'admin_init', array( __CLASS__, 'register_single_site_force_2fa_options' ) );
+		}
+	}
+
+	/**
+	 * Register scripts.
+	 */
+	public static function register_scripts() {
+		// Script for handling AJAX submission in force 2fa takeover screen.
+		wp_register_script(
+			'two-factor-form-script',
+			plugins_url( 'assets/js/force-2fa.js', __FILE__ ),
+			[],
+			'0.1',
+			false
+		);
+	}
+
+	/**
+	 * Redirect a user to the 2fa login screen with redirect parameters.
+	 *
+	 * If a user must have 2fa enabled, we need to send them to the 2fa settings
+	 * takeover. However, we also need to pass in the redirect_to information to
+	 * ensure that the user lands in the right place.
+	 *
+	 * @param string           $redirect_to           The redirect destination URL.
+	 * @param string           $requested_redirect_to The requested redirect destination URL passed as a parameter.
+	 * @param WP_User|WP_Error $user                  WP_User object if login was successful, WP_Error object otherwise.
+	 * @return string
+	 */
+	public static function maybe_redirect_login( $redirect_to, $requested_redirect_to, $user ) {
+		// If login has failed, do nothing.
+		if ( $user instanceof WP_Error ) {
+			return $redirect_to;
+		}
+
+		// Check if redirect is necessary for user.
+		if ( ! self::should_user_redirect( $user->ID ) ) {
+			return $redirect_to;
+		};
+
+		// Append redirect_to URL.
+		return add_query_arg(
+			[
+				'force_2fa_screen' => 1,
+				'redirect_to'      => rawurlencode( $requested_redirect_to ),
+			],
+			admin_url()
+		);
+	}
+
+	/**
+	 * Maybe force the 2fa login page on a user.
+	 *
+	 * If 2fa is required for a user (based on universal or role settings),
+	 * we display the 2-factor options page so that a user must validly enable
+	 * a 2-factor authentication of some kind to perform any action on their site.
+	 * This occurs both on the front and backend.
+	 */
+	public static function maybe_redirect_to_2fa_settings() {
+		if ( ! self::should_user_redirect( get_current_user_id() ) || isset( $_GET['force_2fa_screen'] ) ) {
```

Comment:
Looks like `force_2fa_screen` is re-used in a few places so we could introduce a constant for the key.